Vectorize emulator results handling

alecandido opened the pull request:
Small recap:

1. I replaced the usage of `options.updates`, since those already present are also already consumed at `Platform` level
    https://github.com/qiboteam/qibolab/blob/0fcc6c1e4d4b67ba4977aa4ea6e945156be71aa8/src/qibolab/_core/platform/platform.py#L281
    so, I simply used a standalone `updates` dictionary
2. the core of the proposal was to avoid propagating dictionaries from right after `mesolve` up to the entire collection of all swept values; I replaced them with plain n-dimensional arrays, in which one dimension is dedicated to the measurements, whose order is intended to be the one established by `self._measurements`, and consistent between the samples' extraction, in which I use the values, and the final dictionary reconstruction, in which I establish the output keys from `self._measurements` keys
3. I introduced a plain test, just to ease the purpose of testing - then I realized I needed at least a single meaningful sweeper to check all the reshapes, and I used the following
	```py
	import os
	from pathlib import Path
	
	from qibocal.cli.report import report
	from qibocal.auto.execute import Executor
	
	path = ...
	
	os.environ["QIBOLAB_PLATFORMS"] = str(
	    Path(__file__).parent / "qibolab" / "tests" / "instruments"
	)
	
	with Executor.open("myexec", platform="emulator", path=path, force=True) as e:
	    rs = e.rabi_amplitude(min_amp=0.0, max_amp=1, step_amp=0.01, nshots=1e4)
	
	
	report(path)
	```
    however, it would be nice (and fairly feasible) to extract the sweeper definition from the Qibocal routines, and introduce as a test in Qibolab (more complex tests are not that relevant, since they will boil down to emulator's usage - and we'll do a lot of that, but with Qibocal itself)

There is still material to be polished, especially the `_play` function, for which I should find better names (starting from the function itself, and continuing with all the `res` variables...), and describe a little better what the various steps are doing.

In any case, it is unlikely that this will considerably speed up the execution, since the most expensive part of each loop is still the `mesolve()` invocation. And I fear it is difficult to impossible to vectorize it over sweeper values...
But it should at least streamline a bit the execution, keeping all the merge/reshape complexity at the higher-most level, in a more modular fashion.

alecandido commented:
Btw, I did not touch the `EmulatorController._measurement()` and `._pulse_sequence_to_hamiltonian()` functions, working mostly on updates and results.

I intend to properly review even those, but not in this PR.

andrea-pasquale reviewed (commented):
Thanks @alecandido.
Before I check the code I tried to run the example that you proposed in the comment above. However the code is failing with the following error:
<details closed>
<summary>ERROR</summary>

```bash
[Qibo 0.2.16|INFO|2025-03-10 09:16:48]: Loading platform dummy
[Qibo 0.2.16|INFO|2025-03-10 09:16:48]: Loading platform dummy
[Qibo 0.2.16|INFO|2025-03-10 09:16:48]: Loading platform emulator
[Qibo 0.2.16|INFO|2025-03-10 09:16:48]: Loading platform emulator
[Qibo 0.2.16|INFO|2025-03-10 09:16:48]: Loading platform emulator
[Qibo 0.2.16|INFO|2025-03-10 09:16:48]: Loading platform emulator
[Qibo 0.2.16|INFO|2025-03-10 09:16:48]: Loading platform emulator
[Qibocal 0.1.3|WARNING|2025-03-10 09:16:48]: Deleting previous directory test_emulator.
[Qibocal 0.1.3|INFO|2025-03-10 09:16:48]: Creating directory test_emulator.
[Qibocal 0.1.3|INFO|2025-03-10 09:16:48]: Executing mode AUTOCALIBRATION on rabi_amplitude.
[Qibo 0.2.16|INFO|2025-03-10 09:16:48]: Minimal execution time: 2.1
Traceback (most recent call last):
  File "/Users/andreapasquale/qibolab/test_emulator.py", line 14, in <module>
    rs = e.rabi_amplitude(min_amp=0, max_amp=1, step_amp=0.01, nshots=1e4)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/andreapasquale/qibocal/src/qibocal/auto/execute.py", line 223, in wrapper
    return self.run_protocol(protocol, parameters=action, mode=mode)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/andreapasquale/qibocal/src/qibocal/auto/execute.py", line 128, in run_protocol
    completed = task.run(platform=self.platform, targets=self.targets, mode=mode)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/andreapasquale/qibocal/src/qibocal/auto/task.py", line 159, in run
    completed.data, completed.data_time = operation.acquisition(
                                          ^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/andreapasquale/qibocal/src/qibocal/auto/operation.py", line 40, in wrapper
    out = func(*args, **kwds)
          ^^^^^^^^^^^^^^^^^^^
  File "/Users/andreapasquale/qibocal/src/qibocal/protocols/rabi/amplitude.py", line 72, in _acquisition
    results = platform.execute(
              ^^^^^^^^^^^^^^^^^
  File "/Users/andreapasquale/qibolab/src/qibolab/_core/platform/platform.py", line 294, in execute
    results |= self._execute(b, options, configs, sweepers)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/andreapasquale/qibolab/src/qibolab/_core/platform/platform.py", line 228, in _execute
    new_result = instrument.play(configs, sequences, options, sweepers)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/andreapasquale/qibolab/src/qibolab/_core/instruments/emulator/emulator.py", line 125, in play
    return reduce(
           ^^^^^^^
  File "/Users/andreapasquale/qibolab/src/qibolab/_core/instruments/emulator/emulator.py", line 128, in <genexpr>
    self._play(sequence, configs, options, sweepers)
  File "/Users/andreapasquale/qibolab/src/qibolab/_core/instruments/emulator/emulator.py", line 98, in _play
    res = self._sweep(sequence, configs, sweepers)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/andreapasquale/qibolab/src/qibolab/_core/instruments/emulator/emulator.py", line 87, in _sweep
    results.append(self._sweep(sequence, configs, sweepers[1:], updates))
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/andreapasquale/qibolab/src/qibolab/_core/instruments/emulator/emulator.py", line 76, in _sweep
    return self._play_sequence(sequence, configs, updates)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/andreapasquale/qibolab/src/qibolab/_core/instruments/emulator/emulator.py", line 60, in _play_sequence
    results = mesolve(
              ^^^^^^^^
  File "/Users/andreapasquale/default/lib/python3.12/site-packages/qutip/solver/mesolve.py", line 151, in mesolve
    H = QobjEvo(H, args=args, tlist=tlist)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "qutip/core/cy/qobjevo.pyx", line 223, in qutip.core.cy.qobjevo.QobjEvo.__init__
  File "qutip/core/cy/qobjevo.pyx", line 272, in qutip.core.cy.qobjevo.QobjEvo._read_element
  File "/Users/andreapasquale/default/lib/python3.12/site-packages/qutip/core/coefficient.py", line 195, in coefficient
    raise TypeError("The coefficient function must return a number")
TypeError: The coefficient function must return a number
```
</details>
This particular example also fails on the `emulator` branch (using https://github.com/qiboteam/qibocal/pull/1103 with the platform `emulator1q`) but the error is `ValueError: probabilities are not non-negative` which leads me to think that the probabilities are not defined properly with amplitude set to 0. By changing the parameters the minimum amplitude to `0.1` I am getting some results.

![Screenshot 2025-03-10 at 09 28 57](https://github.com/user-attachments/assets/d428a95a-3803-4d14-837c-9b62f441856e)

Here is the script which I have used: 

<details closed>
<summary>SCRIPT</summary>

```py
import os
from pathlib import Path

from qibocal.cli.report import report
from qibocal.auto.execute import Executor

path = Path("test_emulator")

# os.environ["QIBOLAB_PLATFORMS"] = str(
#     Path(__file__).parent.parent / "qibolab" / "tests" / "instruments"
# )

with Executor.open("myexec", platform="emulator1q", path=path, force=True) as e:
    rs = e.rabi_amplitude(min_amp=0.1, max_amp=0.4, step_amp=0.01, nshots=1e4)


report(path)
```
</details>
Using the same "fix" on this branch I am getting this error instead 

<details closed>

<summary>ERROR</summary>

```bash
[Qibo 0.2.16|INFO|2025-03-10 09:27:35]: Loading platform dummy
[Qibo 0.2.16|INFO|2025-03-10 09:27:35]: Loading platform dummy
[Qibo 0.2.16|INFO|2025-03-10 09:27:35]: Loading platform emulator1q
[Qibo 0.2.16|INFO|2025-03-10 09:27:35]: Loading platform emulator1q
[Qibo 0.2.16|INFO|2025-03-10 09:27:35]: Loading platform emulator1q
[Qibo 0.2.16|INFO|2025-03-10 09:27:35]: Loading platform emulator1q
[Qibo 0.2.16|INFO|2025-03-10 09:27:35]: Loading platform emulator1q
[Qibocal 0.1.3|WARNING|2025-03-10 09:27:35]: Deleting previous directory test_emulator.
[Qibocal 0.1.3|INFO|2025-03-10 09:27:35]: Creating directory test_emulator.
[Qibocal 0.1.3|INFO|2025-03-10 09:27:35]: Executing mode AUTOCALIBRATION on rabi_amplitude.
[Qibo 0.2.16|INFO|2025-03-10 09:27:35]: Minimal execution time: 0.651
[Qibocal 0.1.3|INFO|2025-03-10 09:27:40]: Finished acquisition in 4.78 seconds
[Qibocal 0.1.3|INFO|2025-03-10 09:27:40]: Finished fit in less than 1 second.
(default) andreapasquale ~/qibolab [emulator] $ open test_emulator/index.html 
(default) andreapasquale ~/qibolab [emulator] $ git ch vectorize-emulator
Switched to branch 'vectorize-emulator'
Your branch is up to date with 'origin/vectorize-emulator'.
(default) andreapasquale ~/qibolab [vectorize-emulator] $ python test_emulator.py      
[Qibo 0.2.16|INFO|2025-03-10 09:35:15]: Loading platform dummy
[Qibo 0.2.16|INFO|2025-03-10 09:35:15]: Loading platform dummy
Traceback (most recent call last):
  File "/Users/andreapasquale/qibolab/test_emulator.py", line 13, in <module>
    with Executor.open("myexec", platform="emulator", path=path, force=True) as e:
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/contextlib.py", line 137, in __enter__
    return next(self.gen)
           ^^^^^^^^^^^^^^
  File "/Users/andreapasquale/qibocal/src/qibocal/auto/execute.py", line 320, in open
    ex = cls.create(name, platform)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/andreapasquale/qibocal/src/qibocal/auto/execute.py", line 106, in create
    else create_calibration_platform(
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/andreapasquale/qibocal/src/qibocal/calibration/platform.py", line 37, in create_calibration_platform
    platform = create_platform(name)
               ^^^^^^^^^^^^^^^^^^^^^
  File "/Users/andreapasquale/qibolab/src/qibolab/_core/platform/load.py", line 81, in create_platform
    hardware = _load(path)
               ^^^^^^^^^^^
  File "/Users/andreapasquale/qibolab/src/qibolab/_core/platform/load.py", line 48, in _load
    return module.create()
           ^^^^^^^^^^^^^^^
  File "/Users/andreapasquale/qibolab_platforms_qrc/emulator/platform.py", line 32, in create
    return Platform.load(
           ^^^^^^^^^^^^^^
  File "/Users/andreapasquale/qibolab/src/qibolab/_core/platform/platform.py", line 308, in load
    parameters = Parameters.model_validate_json((path / PARAMETERS).read_text())
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/andreapasquale/default/lib/python3.12/site-packages/pydantic/main.py", line 656, in model_validate_json
    return cls.__pydantic_validator__.validate_json(json_data, strict=strict, context=context)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
pydantic_core._pydantic_core.ValidationError: 1 validation error for Parameters
configs.hamiltonian.single_qubit.0.anharmonicity
  Extra inputs are not permitted [type=extra_forbidden, input_value=-200000000, input_type=int]
    For further information visit https://errors.pydantic.dev/2.10/v/extra_forbidden
(default) andreapasquale ~/qibolab [vectorize-emulator] $ python test_emulator.py
[Qibo 0.2.16|INFO|2025-03-10 09:35:26]: Loading platform dummy
[Qibo 0.2.16|INFO|2025-03-10 09:35:26]: Loading platform dummy
[Qibo 0.2.16|INFO|2025-03-10 09:35:27]: Loading platform emulator
[Qibo 0.2.16|INFO|2025-03-10 09:35:27]: Loading platform emulator
[Qibo 0.2.16|INFO|2025-03-10 09:35:27]: Loading platform emulator
[Qibo 0.2.16|INFO|2025-03-10 09:35:27]: Loading platform emulator
[Qibo 0.2.16|INFO|2025-03-10 09:35:27]: Loading platform emulator
[Qibocal 0.1.3|WARNING|2025-03-10 09:35:27]: Deleting previous directory test_emulator.
[Qibocal 0.1.3|INFO|2025-03-10 09:35:27]: Creating directory test_emulator.
[Qibocal 0.1.3|INFO|2025-03-10 09:35:27]: Executing mode AUTOCALIBRATION on rabi_amplitude.
[Qibo 0.2.16|INFO|2025-03-10 09:35:27]: Minimal execution time: 0.651
Traceback (most recent call last):
  File "/Users/andreapasquale/qibolab/test_emulator.py", line 14, in <module>
    rs = e.rabi_amplitude(min_amp=0.1, max_amp=0.4, step_amp=0.01, nshots=1e4)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/andreapasquale/qibocal/src/qibocal/auto/execute.py", line 223, in wrapper
    return self.run_protocol(protocol, parameters=action, mode=mode)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/andreapasquale/qibocal/src/qibocal/auto/execute.py", line 128, in run_protocol
    completed = task.run(platform=self.platform, targets=self.targets, mode=mode)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/andreapasquale/qibocal/src/qibocal/auto/task.py", line 159, in run
    completed.data, completed.data_time = operation.acquisition(
                                          ^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/andreapasquale/qibocal/src/qibocal/auto/operation.py", line 40, in wrapper
    out = func(*args, **kwds)
          ^^^^^^^^^^^^^^^^^^^
  File "/Users/andreapasquale/qibocal/src/qibocal/protocols/rabi/amplitude.py", line 72, in _acquisition
    results = platform.execute(
              ^^^^^^^^^^^^^^^^^
  File "/Users/andreapasquale/qibolab/src/qibolab/_core/platform/platform.py", line 294, in execute
    results |= self._execute(b, options, configs, sweepers)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/andreapasquale/qibolab/src/qibolab/_core/platform/platform.py", line 228, in _execute
    new_result = instrument.play(configs, sequences, options, sweepers)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/andreapasquale/qibolab/src/qibolab/_core/instruments/emulator/emulator.py", line 125, in play
    return reduce(
           ^^^^^^^
  File "/Users/andreapasquale/qibolab/src/qibolab/_core/instruments/emulator/emulator.py", line 128, in <genexpr>
    self._play(sequence, configs, options, sweepers)
  File "/Users/andreapasquale/qibolab/src/qibolab/_core/instruments/emulator/emulator.py", line 98, in _play
    res = self._sweep(sequence, configs, sweepers)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/andreapasquale/qibolab/src/qibolab/_core/instruments/emulator/emulator.py", line 87, in _sweep
    results.append(self._sweep(sequence, configs, sweepers[1:], updates))
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/andreapasquale/qibolab/src/qibolab/_core/instruments/emulator/emulator.py", line 76, in _sweep
    return self._play_sequence(sequence, configs, updates)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/andreapasquale/qibolab/src/qibolab/_core/instruments/emulator/emulator.py", line 60, in _play_sequence
    results = mesolve(
              ^^^^^^^^
  File "/Users/andreapasquale/default/lib/python3.12/site-packages/qutip/solver/mesolve.py", line 151, in mesolve
    H = QobjEvo(H, args=args, tlist=tlist)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "qutip/core/cy/qobjevo.pyx", line 223, in qutip.core.cy.qobjevo.QobjEvo.__init__
  File "qutip/core/cy/qobjevo.pyx", line 272, in qutip.core.cy.qobjevo.QobjEvo._read_element
  File "/Users/andreapasquale/default/lib/python3.12/site-packages/qutip/core/coefficient.py", line 194, in coefficient
    if not isinstance(op(0), numbers.Number):
                      ^^^^^
  File "qutip/core/cy/coefficient.pyx", line 220, in qutip.core.cy.coefficient.FunctionCoefficient.__call__
  File "/Users/andreapasquale/qibolab/src/qibolab/_core/instruments/emulator/emulator.py", line 181, in time
    return pulse(t, index)
           ^^^^^^^^^^^^^^^
  File "/Users/andreapasquale/qibolab/src/qibolab/_core/instruments/emulator/hamiltonians.py", line 69, in __call__
    i, q = self.envelopes
           ^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/functools.py", line 993, in __get__
    val = self.func(instance)
          ^^^^^^^^^^^^^^^^^^^
  File "/Users/andreapasquale/qibolab/src/qibolab/_core/instruments/emulator/hamiltonians.py", line 56, in envelopes
    return self.pulse.envelopes(self.sampling_rate)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/andreapasquale/qibolab/src/qibolab/_core/pulses/pulse.py", line 69, in envelopes
    return np.array([self.i(sampling_rate), self.q(sampling_rate)])
                     ^^^^^^^^^^^^^^^^^^^^^
  File "/Users/andreapasquale/qibolab/src/qibolab/_core/pulses/pulse.py", line 60, in i
    return self.amplitude * self.envelope.i(samples)
           ~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~
ValueError: operands could not be broadcast together with shapes (31,) (100,) 
```
</details>

alecandido commented:
Thanks @andrea-pasquale for testing this.

Indeed, in order to run, even on the https://github.com/qiboteam/qibolab/tree/emulator branch, I actually patched the code slightly, in a way which I did not report here.
I.e. I just put an `abs()` around `prob`, since I assumed the negative values were originated by tiny fluctuations around zero values.

In any case, it was just a quick fix to run, so I did not publish. But I obtained results with both branches, and I compared them to be the same.

At this point, I will try to reinstall from scratch, and attempt to reproduce one of the results (either your error, or the observed reports).

alecandido commented:
@andrea-pasquale, apparently, the error was introduced in the last commit I added (before requesting the review, i.e. c7001ae, not the current last, which is the fix), in which I introduced the support for parallel sweepers. But, it seems that I did not test this last part... my bad...

However, now I tried again, and it seems to work, with both your parameters and my original ones.
| Restricted | Full scan |
|--------|--------|
| ![image](https://github.com/user-attachments/assets/817f2039-a623-4569-896a-ee2df1a782c4) | ![image](https://github.com/user-attachments/assets/a66a140e-dd26-4067-9777-ed187c63d991) |

andrea-pasquale reviewed (commented):
I can confirm that now the code is working also from my side.
It seems to be also relatively faster, although readability-wise I find a bit more difficult to follow.
Just one comment related to this
> 1. I replaced the usage of `options.updates`, since those already present are also already consumed at `Platform` level
>        https://github.com/qiboteam/qibolab/blob/0fcc6c1e4d4b67ba4977aa4ea6e945156be71aa8/src/qibolab/_core/platform/platform.py#L281

Sooner or later, when we introduce different `ExecutionParameters` we will need to pass the `options` as well, so maybe we could consider going back to the previous implementation at some point (just to avoid having an extra argument). Although on the other hand I undestand that I was perhaps misinterpreting the usage of `options.updates`.

alecandido commented on tests/instruments/emulator/calibration.json:
I needed because I used this platform offline to run some Qibocal routines as well.

On the one hand, I perfectly agree with you, it'd be more minimal without.
On the other: is it optional for Qibocal? I did not check. If it is not, it may be practical to keep it, just to quickly allow running routines as well, though it will never be used other than for offline debugging.

As soon as we implement suitable tests with sweepers, we certainly do not need it any longer (since we can avoid relying on the plot to check that everything works as expected).

andrea-pasquale replied:
For qibocal it is mandatory (especially because some of the routines are actively using parameters from this file).
For debugging I've added an emulator directly in https://github.com/qiboteam/qibocal/pull/1103.
More generally I intend to populate [`platforms`](https://github.com/qiboteam/qibocal/tree/emulator/platforms) with exactly these "debugging platforms" which in this case are just `mock` and `emulator1q`.

alecandido replied:
I'm thinking whether to ask to keep this file temporarily, just for the convenience of having it in a single repo.

However, it is true that I can just point my `QIBOLAB_PLATFORMS` somewhere else, and load the `emulator1q` as well (of which this is just a copy, for the convenience of the CI).

So, I guess we can drop this file, and just avoid relying on the same tools of Pytest to manually run routines (which I avoided here, and I'd never include in this repo, to prevent any kind of loop).

alecandido commented on src/qibolab/_core/instruments/emulator/emulator.py:
```diff
@@ -46,99 +49,100 @@ def initial_state(self):
+            # update all parallel sweepers with the respective values
+            for sweeper, value in zip(parsweep, values):
+                if sweeper.pulses is not None:
+                    for pulse in sweeper.pulses:
```
I guess you meant the one for *channels*. In which case, it was indeed the bug observed in the 2D Rabi, which it was fixed in bfad4e5

alecandido commented:
@andrea-pasquale: now I consider this to be completed. If you do not have any further suggestion, or anything else to amend, I'd rather merge this PR, and move forward with next enhancements (starting from the simple integration mode, but also flux and so on).
